### PR TITLE
Updated model interface

### PR DESCRIPTION
**Pull request by xukai92**
This PR implements the new interface described in #82.

Summary of PR:

1. Allow declaring parameters in model macro;
2. Support = in model declaration;
    - The old model declaration syntax is still support
3. Use `@sample` macro to run inference engine.
    - This macro construct the `data` dictionary in the previous design and run the `sample()` function.
    - Calling `@sample` is the same as constructing the `data` dictionary and call `sample(model, data=..., alg)`.

### CODE REVIEW

**Comment by yebai** [comment]
@xukai92 It seems some temporary stan files (see tmp/) are also included in the recent commit. Could you exclude them?

**Comment by xukai92** [comment]
@yebai Done.

File: src/samplers/support/hmc_helper.jl [hunk 1]
```diff
@@ -25,7 +25,12 @@ end
 end
 
 Base.convert{N,T<:Real}(::Type{T}, d::Dual{N,T})  = d.value
-Base.convert(::Type{Float64}, d::Dual{0,Int}) = round(Int, d.value)
+Base.convert{N}(::Type{Int}, d::Dual{N,Float64}) = round(Int, d.value)
```
Comment (tkelman) [inline src/samplers/support/hmc_helper.jl]: Do you need to be extending the conversion methods of a different package's type here? That can change the behavior of unrelated code

Comment (xukai92) [inline src/samplers/support/hmc_helper.jl]: This is removed now.
